src/v/config/constraints.cc

```diff
@@ -45,6 +54,534 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+  const std::optional<T>& cluster_val,
+  const constraint_enabled_t& enabled,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property) {
+    if (enabled == constraint_enabled_t::no) {
+        // Since the constraint is turned off, there is no need to clamp
+        return;
+    }
+
+    // A constraint with "enabled" flag means that the topic property gets the
+    // cluster value, but an undefined cluster value cannot be assigned so
+    // ignore.
+    if (!cluster_val) {
+        return;
+    }
+
+    vlog(
+      constraints_log.warn,
+      "Overwriting topic property to the cluster property {}: topic property "
+      "{}.{}, "
+      "value {}",
+      cluster_property,
+      topic(),
+      topic_property,
+      topic_val);
+    topic_val = tristate<T>{cluster_val};
+}
+
+/**
+ * Assigns the constraint range or cluster-level value to the topic-level value
+ * \param topic_val: the topic value
+ * \param constraint: the constraint to evaluate
+ * \param property_name: name of the cluster-level property
+ * \param cluster_opt: the value from the cluster-level property
+ */
+template<typename T>
+void clamp_value(
+  tristate<T>& topic_val,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (constraint.name == cluster_property) {
+        ss::visit(
+          constraint.flags,
+          [&topic_val,
+           &topic,
+           &topic_property,
+           &cluster_property,
+           &cluster_opt](const constraint_enabled_t enabled) {
+              cluster_property_clamp(
+                topic_val,
+                cluster_opt,
+                enabled,
+                topic,
+                topic_property,
+                cluster_property);
+          },
+          [&topic_val, &topic, &topic_property](const auto range) {
+              range_clamp(topic_val, range, topic, topic_property);
+          });
+    }
+}
+
+template<typename T>
+void clamp_value(
+  std::optional<T>& topic_opt,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    auto tri = tristate<T>{topic_opt};
+    clamp_value(
+      tri, constraint, topic, topic_property, cluster_property, cluster_opt);
+    // The tristate is not in disabled state since it was assigned an optional
+    // earlier.
+    topic_opt = tri.get_optional();
+}
+} // namespace
+
+void constraint_clamp_topic_config(
+  cluster::topic_configuration& topic_cfg, const constraint_t& constraint) {
+    auto partition_count_tri = tristate(
+      std::make_optional(topic_cfg.partition_count));
+    auto replication_factor_tri = tristate(
+      std::make_optional(topic_cfg.replication_factor));
```

Comment:
It's not obvious to me why these ones are separate. I see a few cases:
* The topic property is a tristate, it's taken by reference and modified
* The topic property is an optional, an intermediate tristate is constructed, and the optional is written back.
* The topic property is is remote_delete - it's ignored
* Oh, I see, these are topic configs!
  * Is it worth hiding these write-backs inside a clamp_value that takes a scaler and wraps it in an optional, ensuring it's written back?